perma_web/fabfile/dev.py

```diff
@@ -573,3 +573,113 @@ def check_s3_hashes():
                 m.update(buf)
         if m.hexdigest() != remote_paths[local_path]:
             print "Hash mismatch! Local: %s Remote: %s" % (m.hexdigest(), remote_paths[local_path])
+
+
+@task
+def check_storage():
+    """
+        Confirm that, for every link, there is a WARC in each storage, and that their hashes match.
+
+        Ground truth is the list of link objects: compare its list of warcs with those of each storage,
+        and compare hashes when more than one such file is present.
+
+        Derived from check_s3_hashes
+    """
+    from django.core.files.storage import default_storage
+    from django.db.models import Q
+    from perma.models import Link, Capture
+    from tqdm import tqdm
+
+    # this can be generalized later to an arbitrary number of storages
+    storages = {'primary': {'storage': default_storage, 'lookup': {}}}
+    if hasattr(default_storage, 'secondary_storage'):
+        storages.update({'secondary': {'storage': default_storage.secondary_storage, 'lookup': {}}})
+
+    # only use cache files when all are present: link cache, and one for each storage
+    link_cache = '/tmp/perma_link_cache.txt'
+    caches = [link_cache]
+    for key in storages:
+        caches.append('/tmp/perma_storage_cache_{0}.txt'.format(key))
+
+    if not all(map(os.path.exists, caches)):
```

Comment:
Hooray for all and map! Since we are talking about a list of length 2 at this point, certainly no optimization is needed, but mentioning just for fun: it might be preferable to use a generator here instead of map. ```all(os.path.exists(p) for p in caches)``` would stop as soon as os.path.exists returned false; unless I'm mistaken, all(map...) will calculate os.path.exists for every p in caches, and then start looking for a falsy value. But again, totally doesn't matter here.